Groups pages table

liyiy opened the pull request:
## Changes

*Please describe.*  

Adding support for groups (like persons table)

https://user-images.githubusercontent.com/25164963/141602976-28c0b2b5-c250-46cb-9204-18e5b5080896.mov


*If this affects the frontend, include screenshots.*  

## How did you test this code?

*Please describe.*

liyiy committed:
wip

liyiy committed:
get list of groups

liyiy committed:
Merge branch 'master' into groups-pages

liyiy committed:
move backend endpoints

liyiy committed:
frontend wip

liyiy committed:
fix routes for groups

liyiy committed:
add groups file

liyiy committed:
try to fix tab nav

liyiy committed:
fix groups query

liyiy committed:
reformat results, add limits and offsets and next url

liyiy committed:
move groups pages logic into own file

liyiy committed:
fix and clean up tabs

liyiy committed:
account for pagination and response reformatting

liyiy commented on ee/clickhouse/views/groups.py:
```diff
@@ -18,11 +18,49 @@ class Meta:
 
 
-class ClickhouseGroupsView(StructuredViewSetMixin, ListModelMixin, viewsets.GenericViewSet):
+class ClickhouseGroupsTypesView(StructuredViewSetMixin, ListModelMixin, viewsets.GenericViewSet):
```
is it okay to leave this types view in this file since it's so small?

macobo replied:
Yup, they are closely related

liyiy committed:
lint fixes

liyiy committed:
optional prev input

liyiy committed:
typescript errors

liyiy committed:
typescript fixes

liyiy commented:
So my localhost broke while testing this so I will update tomorrow morning with screenshots/video, but so far we have the basic table implementation. I still have to add the properties expanded row but other than that this should be good for review

mariusandra commented:
I'm just your flyby reminder that every logic change, including new logics, [should come with tests](https://posthog.com/docs/contribute/coding-conventions#frontend).

macobo commented on ee/clickhouse/views/groups.py:
```diff
@@ -18,11 +18,49 @@ class Meta:
+    pagination_class = None
+
+    def list(self, request, *args, **kwargs):
+        is_csv_request = self.request.accepted_renderer.format == "csv"
```
Drop the is_csv behavior for now - we can add it later as needed :)

macobo commented on posthog/api/utils.py:
```diff
@@ -23,19 +23,19 @@ def get_target_entity(request: request.Request) -> Entity:
 
 
-def format_next_url(request: request.Request, offset: int, page_size: int):
+def format_next_url(request: request.Request, offset: int, page_size: int, previous: Optional[bool] = None):
```
`Optional[bool]` doesn't really make sense - you want `previous = False` here instead.

Also the method name is now confusing. Is there a better way to name this function given it handles previous page as well?

liyiy replied:
`format_paginate_url`?

macobo replied:
`format_paginated_url` sounds even better

macobo commented on frontend/src/scenes/groups/Groups.tsx:
```diff
@@ -0,0 +1,72 @@
+                        disabled={!groups.next_url}
+                        onClick={() => {
+                            loadGroups(groups.next_url)
+                            window.scrollTo(0, 0)
```
Q: Why do we need this?

macobo reviewed (commented):
Please add images/gifs/explanations to your flow into the PR itself - this makes it easier to understand the flow on a high level.

macobo commented on frontend/src/scenes/groups/Groups.tsx:
```diff
@@ -0,0 +1,72 @@
+                dataSource={groups.results}
+            />
+            {(groups.previous_url || groups.next_url) && (
+                <div style={{ margin: '3rem auto 10rem', width: 200, display: 'flex', alignItems: 'center' }}>
```
Let's write a component-specific SCSS file instead and assign this a class.

Given this is really similar with persons, perhaps we can share styles?

macobo commented on frontend/src/scenes/groups/groupsListLogic.ts:
```diff
@@ -0,0 +1,64 @@
+    connect: { values: [teamLogic, ['currentTeamId'], groupsModel, ['groupsEnabled', 'groupTypes']] },
+    actions: () => ({
+        loadGroups: (url?: string | null) => ({ url }),
+        setTab: (tab: string) => ({ tab }),
```
Why is tab a string here and a number below? This causes you headaches with typecasting.

macobo commented on frontend/src/scenes/persons/Persons.tsx:
```diff
@@ -36,7 +39,8 @@ export function Persons({ cohort }: PersonsProps = {}): JSX.Element {
     return (
         <div className="persons-list">
-            {!cohort && <PageHeader title="Persons" />}
+            {!cohort && !groupsEnabled && <PageHeader title="Persons" />}
```
Use a ternary here instead:

```
!cohort && (groupsEnabled ? <PageHeader /> : <GroupsTabs />
```

macobo commented on frontend/src/scenes/groups/groupsListLogic.ts:
```diff
@@ -0,0 +1,64 @@
+                    if (values.groupsEnabled) {
+                        if (!url) {
+                            return await api.get(
+                                `api/projects/${values.currentTeamId}/groups/?group_type_index=${values.currentGroup}`
```
You're repeating yourself here.

```
url = url || `api/projects/${values.currentTeamId}/groups/?group_type_index=${values.currentGroup}`
```

macobo commented on frontend/src/scenes/groups/groupsListLogic.ts:
```diff
@@ -0,0 +1,64 @@
+                        }
+                        return await api.get(url)
+                    }
+                    return []
```
This does not match the structure of `GroupsPaginatedResponse`.

macobo reviewed (commented):
Copy-pasta, ignore.

liyiy committed:
address feedback comments

liyiy committed:
groups list logic test

liyiy committed:
render properties expandable row

liyiy committed:
fix test and lints

liyiy committed:
prettier

liyiy committed:
simplify table column title

macobo committed:
Fix a typing issue

macobo committed:
Use enum > boolean parameter

This makes what's going on more explicit

macobo committed:
Turn off pagination at the bottom of the table

macobo committed:
Rename table column

macobo committed:
Make pagination in groups work

Quite bugs fixed:
1. `previous` page was always shown, even if going into negative offsets
2. query_result could never be larger than `limit` with the code

macobo committed:
Cleanup code

macobo committed:
Make sidenav active under persons

macobo commented on ee/clickhouse/views/groups.py:
```diff
@@ -18,11 +18,48 @@ class Meta:
+                    WHERE team_id = %(team_id)s AND group_type_index = %(group_type_index)s
+                    GROUP BY group_key
+                    LIMIT 100 OFFSET %(offset)s
+                ) latest_groups
```
The query approach here is wrong. As I mentioned in the latest pairing session, the approach is `group by` and 
 `argMax` instead.

macobo committed:
Use the correct query

macobo committed:
Add tests for pagination

macobo committed:
Update wording again

macobo committed:
Merge remote-tracking branch 'origin/master' into groups-pages

macobo committed:
Merge remote-tracking branch 'origin/master' into groups-pages

macobo commented:
Merging this in. Fixed couple of pretty large problems:
1. The query was incorrect and didn't work
2. Pagination did not work 
3. Double pagination was being shown on the frontend (one provided by the Table element)
4. No backend tests

Suggested topic for next pairing session: Creating great PR descriptions and making PRs easy to review.
